compiler/repl-service/server/src/main/scala/com/digitalasset/daml/lf/ReplServiceMain.scala

```diff
@@ -187,12 +206,15 @@ class ReplService(val clients: Participants[LedgerClient], ec: ExecutionContext,
       .decoder
     val lfScenarioModule =
       dop.protoScenarioModule(Decode.damlLfCodedInputStream(req.getDamlLf1.newInput))
-    val mod: Ast.Module = dop.decodeScenarioModule(homePackageId, lfScenarioModule)
+    // ConcurrentCompiledPackages doesn't allow overwriting a previously loaded
```

Comment:
This seems a bit problematic. You are creating a memory leak here where each line will create a new LF package and they never get garbage collected. While we already have a memory leak since we never collect results even if they are unused (e.g. they are not bound to a name or are shadowed), this one seems a bit more severe (an LF package is likely to be bigger than most results) and is harder to solve.

Looking at the code, do we actually benefit from `ConcurrentCompiledPackages` over `PureCompiledPackages`? The latter has a constructor that allows you to pass in the compiled expressions (the `apply` method takes care of the compilation otherwise). If we use that we should be able to drop old packages. There is one issue here, namely if the result of some line is a closure as opposed to a simple value, it can keep references to a package. Without this PR that simply fails since we always drop packages. I think with this PR it might actually work?

I’m fine with using `ConcurrentCompiledPackages` for now and then see if we can switch in anther PR. Maybe just update the comment.

Reply:
Hmm, yes, I think we could skip `ConcurrentCompiledPackages` altogether and call the compiler directly within "load package" and use `PureCompiledPackages` for the script runner. That would still avoid recompilation.

Reply:
I've updated the PR to do this. That also makes for a simpler change.

Reply:
Very nice! That looks so simple that I’m almost slightly embarrassed that I didn’t implement this in the beginning :slightly_smiling_face: